fix: prevent too many Validation Rules errors

luciano-fiandesio opened the pull request:
DHIS-7966
This fix prevents a Validation Rule to run on data which is related to a
OU that is not selected by the user.
Without this fix, the rules are executed on incorrect data, which cause
the test expression to always return false.

luciano-fiandesio commented:
> I believe this fix will not work in the following case: Define 2 validation rules in a group. Rule A has no data for a given orgUnit and should not be evaluated there. Rule B has data for that orgUnit. When validating the ValidationRuleGroup containing both Rule A and Rule B, both will be evaluated for the given orgUnit.
> 
> Instead, I believe the fix should be to just remove the following lines in `validateRule()`:
> 
> ```
>     if ( attributeOptionCombos.isEmpty() )
>     {
>         attributeOptionCombos = Sets.newHashSet( context.getDefaultAttributeCombo().getUid() );
>     }
> ```
> 
> This logic was added to validation rules, following a similar logic in predictors. For example, you can define a predictor with `MissingValueStrategy.NEVER_SKIP` to populate a constant value into a dataElement for every orgUnit. If there is no data for any attribute option combination, then the value will be inserted using the default AOC. However for validation rules, this logic is faulty. If there are no data values for an orgUnit, then the validation rule should not be executed. These lines introduced the bug, and removing them should fix it.

Hi @jimgrace, yes, I'm aware that the solution is not what we discussed. 
I have created a unit test that tries to reproduce this specific problem and the test is green. May it be that the test is wrong? The test is: `verifySimpleValidation_twoRules_noErrors`. Can you please double check if I misunderstood the test setup?

jimgrace commented:
Hi @luciano-fiandesio, The test you reference has the same dataElement.catOptionCombo for both rules.

A test that I believe will fail is if Rule A references a de.coc where there is no data for that orgUnit, but Rule B references a different de.coc where there is data for the same orgUnit. As I read the logic, both rules will be evaluated, even though only Rule B should be evaluated for that orgUnit, and Rule A should not.

luciano-fiandesio commented:
> Hi @luciano-fiandesio, The test you reference has the same dataElement.catOptionCombo for both rules.
> 
> A test that I believe will fail is if Rule A references a de.coc where there is no data for that orgUnit, but Rule B references a different de.coc where there is data for the same orgUnit. As I read the logic, both rules will be evaluated, even though only Rule B should be evaluated for that orgUnit, and Rule A should not.

@jimgrace, I have removed the OU check and removed the empty aoc check (as you suggested).
